File: web/src/message_find.ts
export interface SearchHit {
  block: number;
  start: number;
  end: number;
}
export type SearchText = (
  blocks: string[],
  query: string,
  matchCase: boolean,
) => Promise<SearchHit[]>;
export class MessageFind extends EventTarget {
  query = "";
  identity = "";
  blocks: string[] = [];
  hits: SearchHit[] = [];
  open = false;
  matchCase = false;
  pending = false;
  active = 0;
  revision = 0;
  jump = 0;
  error: string | null = null;
  private running = false;
  private ready = false;
  private timer?: ReturnType<typeof setTimeout>;
  private disposed = false;
  constructor(
    private search: SearchText,
    private debounce = 100,
  ) {
    super();
  }
  get status() {
    return this.pending
      ? "Searching…"
      : !this.query
        ? "Search this message"
        : !this.hits.length
          ? "No matches"
          : `${this.active + 1} of ${this.hits.length}`;
  }
  private changed() {
    this.dispatchEvent(new Event("change"));
  }
  setSource(id: string, blocks: string[]) {
    if (
      this.identity === id &&
      blocks.length === this.blocks.length &&
      blocks.every((s, i) => s === this.blocks[i])
    )
      return;
    this.identity = id;
    this.blocks = [...blocks];
    this.invalidate();
  }
  show() {
    this.open = true;
    this.invalidate();
  }
  close() {
    this.open = false;
    this.invalidate();
  }
  setQuery(value: string) {
    this.query = value;
    this.invalidate();
  }
  toggleCase() {
    this.matchCase = !this.matchCase;
    this.invalidate();
  }
  retry() {
    this.invalidate();
  }
  next(previous = false) {
    if (!this.hits.length || this.pending) return;
    this.active =
      (this.active + (previous ? -1 : 1) + this.hits.length) % this.hits.length;
    this.jump++;
    this.changed();
  }
  private invalidate() {
    this.revision++;
    this.hits = [];
    this.active = 0;
    this.error = null;
    clearTimeout(this.timer);
    this.ready = false;
    this.pending = this.open && !!this.query;
    if (this.pending)
      this.timer = setTimeout(() => {
        this.ready = true;
        void this.pump();
      }, this.debounce);
    this.changed();
  }
  private async pump() {
    if (this.disposed || this.running || !this.ready || !this.open) return;
    this.ready = false;
    this.running = true;
    const revision = this.revision;
    try {
      const hits = await this.search(
        [...this.blocks],
        this.query,
        this.matchCase,
      );
      if (!this.disposed && this.open && revision === this.revision) {
        this.hits = hits;
        this.active = 0;
        this.pending = false;
        this.error = null;
        this.jump++;
        this.changed();
      }
    } catch {
      if (!this.disposed && this.open && revision === this.revision) {
        this.pending = false;
        this.error = "Could not search this message. Retry Find.";
        this.changed();
      }
    } finally {
      this.running = false;
      if (!this.disposed && this.ready) void this.pump();
    }
  }
  dispose() {
    this.disposed = true;
    clearTimeout(this.timer);
  }
}
export class SearchWorker {
  private worker?: Worker;
  private loading?: Promise<void>;
  private cancelReady?: () => void;
  private disposed = false;
  private resolve?: (hits: SearchHit[]) => void;
  private reject?: (error: Error) => void;
  constructor() {
    void this.ready().catch(() => {});
  }
  private ready(): Promise<void> {
    if (this.disposed) return Promise.reject(new Error("Find closed."));
    if (this.loading) return this.loading;
    const worker = new Worker(new URL("./search_worker.ts", import.meta.url), {
      type: "module",
    });
    this.worker = worker;
    this.loading = new Promise((resolveReady, rejectReady) => {
      this.cancelReady = () => rejectReady(new Error("Find closed."));
      const failed = () => {
        if (this.worker !== worker) return;
        const error = new Error("Find stopped. Retry the search.");
        rejectReady(error);
        this.reject?.(error);
        this.resolve = this.reject = undefined;
        worker.terminate();
        this.worker = undefined;
        this.loading = undefined;
      };
      worker.onmessage = ({ data }) => {
        if (this.worker !== worker) return;
        if (data.ready) {
          resolveReady();
          return;
        }
        if (data.startup) {
          failed();
          return;
        }
        const resolve = this.resolve,
          reject = this.reject;
        this.resolve = this.reject = undefined;
        if (data.error) reject?.(new Error(data.error));
        else resolve?.(data.hits);
      };
      worker.onerror = failed;
    });
    return this.loading;
  }
  search: SearchText = async (blocks, query, matchCase) => {
    await this.ready();
    if (this.disposed) throw new Error("Find closed.");
    if (this.resolve) throw new Error("Find is busy. Retry shortly.");
    return new Promise((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
      this.worker!.postMessage({ blocks, query, matchCase });
    });
  };
  dispose() {
    this.disposed = true;
    this.cancelReady?.();
    this.worker?.terminate();
    this.reject?.(new Error("Find closed."));
    this.resolve = this.reject = undefined;
  }
}

File: flutter/lib/model/message_find.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import '../data/message_search.dart';

typedef SearchText =
    Future<List<SearchHit>> Function(List<String>, String, bool);

/// One active search and one coalesced latest request. Older results and errors
/// cannot replace a new query, message, quote scope or closed search bar.
class MessageFind extends ChangeNotifier {
  MessageFind(this.search, {this.debounce = const Duration(milliseconds: 100)});
  final SearchText search;
  final Duration debounce;
  String query = '', identity = '';
  List<String> blocks = [];
  List<SearchHit> hits = [];
  bool open = false, matchCase = false, pending = false;
  int active = 0, revision = 0, jump = 0;
  String? error;
  bool _running = false, _ready = false, _disposed = false;
  Timer? _timer;
  String get status => pending
      ? 'Searching…'
      : query.isEmpty
      ? 'Search this message'
      : hits.isEmpty
      ? 'No matches'
      : '${active + 1} of ${hits.length}';
  void setSource(String id, List<String> next) {
    if (identity == id && listEquals(blocks, next)) return;
    identity = id;
    blocks = List.of(next);
    _invalidate();
  }

  void show() {
    open = true;
    _invalidate();
  }

  void close() {
    open = false;
    _invalidate();
  }

  void setQuery(String value) {
    query = value;
    _invalidate();
  }

  void toggleCase() {
    matchCase = !matchCase;
    _invalidate();
  }

  void retry() => _invalidate();
  void next([bool previous = false]) {
    if (hits.isEmpty || pending) return;
    active = (active + (previous ? -1 : 1)) % hits.length;
    jump++;
    notifyListeners();
  }

  void _invalidate() {
    revision++;
    hits = [];
    active = 0;
    error = null;
    _timer?.cancel();
    _ready = false;
    pending = open && query.isNotEmpty;
    if (pending) {
      _timer = Timer(debounce, () {
        _ready = true;
        _pump();
      });
    }
    notifyListeners();
  }

  Future<void> _pump() async {
    if (_disposed || _running || !_ready || !open) return;
    _ready = false;
    _running = true;
    final version = revision;
    try {
      final result = await search(List.of(blocks), query, matchCase);
      if (!_disposed && open && version == revision) {
        hits = result;
        active = 0;
        pending = false;
        error = null;
        jump++;
        notifyListeners();
      }
    } catch (_) {
      if (!_disposed && open && version == revision) {
        pending = false;
        error = 'Could not search this message. Retry Find.';
        notifyListeners();
      }
    } finally {
      _running = false;
      if (!_disposed && _ready) unawaited(_pump());
    }
  }

  @override
  void dispose() {
    _disposed = true;
    _timer?.cancel();
    super.dispose();
  }
}
